daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -62,17 +62,61 @@ private[preprocessing] final class TransactionPreprocessor(
       commands: BackStack[speedy.Command],
   ) {
     def update(
-        newCids: Iterable[ContractId],
+        newInputCids: Iterable[ContractId],
         newLocalCids: Iterable[ContractId],
         cmd: speedy.Command,
-    ): Acc = {
-      val globalCids = this.globalCids ++ newCids.filterNot(localCids)
-      if (newLocalCids.exists(globalCids))
-        fail("Conflicting discriminators between a global and local contract ID.")
-      Acc(globalCids, localCids ++ newLocalCids, commands :+ cmd)
-    }
+    ) = Acc(
+      globalCids ++ newInputCids.filterNot(localCids),
+      localCids ++ newLocalCids,
+      commands :+ cmd,
+    )
   }
 
+  /*
+   * Translates a transaction tree into a sequence of Speedy commands
+   * and collect the global contract IDs.
+   * A contract ID `cid` is considered *local* w.r.t. a node `n`, if
+   * either:
+   *  - it is local in a previous node w.r.t. traversal order, or
+   *  - `n` is a create node such that `n.coid == cid`
+   *
+   * A contract ID `cid` is considered *global* in a root node `n`,
+   * if:
+   *  - `cid` is not considered local w.r.t. `n`, and
+   *  - if `cid` is an input of a `n`, i.e. :
```

Comment:
```suggestion
   *  - if `cid` is referenced in `n`, i.e. :
```
Input contract is typically used for the contract ID that an exercise acts on, not the ids that appear in the choice argument.